File: Python/hello/models/database_functions.py
#!python
# -*- coding: utf-8 -*-

"""Database module. Handles all calls to the User database.

Python version used: 3.6.8
SQLite version used: 3.21.0

Styling guide: PEP 8 -- Style Guide for Python Code
    (https://www.python.org/dev/peps/pep-0008/) and
    PEP 257 -- Docstring Conventions
    (https://www.python.org/dev/peps/pep-0257/)
"""

from __future__ import print_function

import datetime
import os
import sqlite3
import sys

from models import common_functions as co

# Module metadata dunders
__author__ = "Rob Garcia"
__copyright__ = "Copyright 2019-2020, Rob Garcia"
__email__ = "rgarcia@rgprogramming.com"
__license__ = "MIT"

PATH_TO_SQLITE_DB = (co.ROOT_DIR + os.path.sep + "data" + os.path.sep + "users.db")

""" Module functions:
    int create_user_table()
    int create_user(first_name, last_name, email, score, comment)
    list get_all_users()
    tuple get_user_by_user_id(user_id)
    tuple get_user_by_email(email)
    int update_user(user_id, first_name, last_name, email, score, comment)
    int delete_user(user_id)
    int get_next_user_id()
    bool user_exists(email)
    conn connect()
    bool database_exists()
"""


def create_user_table():
    """Creates the User table if it does not exist in the database.

    :returns: The number of rows affected. In Python, both DROP and
        CREATE will return -1.
    :rtype: int
    """
    rows_affected = -1
    try:
        conn = connect()
        cursor = conn.cursor()
        sql = """DROP TABLE IF EXISTS User;"""
        cursor.execute(sql)
        conn.commit()
        sql = """ CREATE TABLE IF NOT EXISTS User (
            UserID integer PRIMARY KEY,
            FirstName text NOT NULL,
            LastName text NOT NULL,
            Email text UNIQUE NOT NULL,
            Score real NOT NULL DEFAULT "100.0",
            CreationDate text NOT NULL,
            Comment text
        );
        """
        cursor.execute(sql)
        conn.commit()
        rows_affected = cursor.rowcount
        cursor.close()
        conn.close()
    except Exception:
        ex = co.log_error(sys.exc_info())
        if co.DISPLAY_ERRORS:
            print(ex)
    return rows_affected


def create_user(first_name, last_name, email, score, comment):
    """Inserts a new user into the database.

    :param first_name: The user's first name.
    :type first_name: str
    :param last_name: The user's last name.
    :type last_name: str
    :param email: The user's email address (can be used as a user name).
    :type email: str
    :param score: The user's score from 0.0 to 100.0.
    :type score: float
    :param comment: Any additional comments.
    :type comment: str
    :returns: The rowid of the new user. A value of 0 indicates an error.
    :rtype: int
    """
    last_row_id = 0
    try:
        # Set other initial values
        user_id = get_next_user_id()
        local_timezone = datetime.datetime.now(
            datetime.timezone.utc).astimezone().tzinfo
        creation_date = datetime.datetime.now(
            local_timezone).strftime("%Y-%m-%d %H:%M:%S")
        conn = connect()
        cursor = conn.cursor()
        sql = "INSERT INTO User VALUES (?, ?, ?, ?, ?, ?, ?);"
        data = (str(user_id), first_name, last_name,
                email, str(score), creation_date, comment)
        cursor.execute(sql, data)
        conn.commit()
        last_row_id = cursor.lastrowid
        cursor.close()
        conn.close()
    except Exception:
        ex = co.log_error(sys.exc_info())
        if co.DISPLAY_ERRORS:
            print(ex)
    return last_row_id


def get_all_users():
    """Gets all the users in the database and their information.

    :returns: A list of all the users in the database and their
        information. An empty list indicates an error.
    :rtype: list
    """
    result = None
    try:
        conn = connect()
        cursor = conn.cursor()
        sql = "SELECT * FROM User ORDER BY UserID ASC;"
        cursor.execute(sql)
        result = cursor.fetchall()
        cursor.close()
        conn.close()
        return result
    except Exception:
        ex = co.log_error(sys.exc_info())
        if co.DISPLAY_ERRORS:
            print(ex)
    return result


def get_user_by_user_id(user_id):
    """Returns a single user and his or her information.

    :param user_id: The user's ID.
    :type user_id: int
    :returns: The user's information indexed by column name or empty
        if the user's ID is not found.
    :rtype: tuple
    """
    result = None
    try:
        conn = connect()
        cursor = conn.cursor()
        sql = "SELECT * FROM User WHERE UserID = ?"
        cursor.execute(sql, str(user_id))
        result = cursor.fetchone()
        cursor.close()
        conn.close()
        return result
    except Exception:
        ex = co.log_error(sys.exc_info())
        if co.DISPLAY_ERRORS:
            print(ex)
    return result


def get_user_by_email(email):
    """Returns a single user and his or her information.

    :param email: The user's email.
    :type email: str
    :returns: The user's information indexed by column name or empty
        if the user's email is not found.
    :rtype: tuple
    """
    result = None
    try:
        conn = connect()
        cursor = conn.cursor()
        sql = "SELECT * FROM User WHERE Email = ?"
        cursor.execute(sql, email)
        result = cursor.fetchone()
        cursor.close()
        conn.close()
        return result
    except Exception:
        ex = co.log_error(sys.exc_info())
        if co.DISPLAY_ERRORS:
            print(ex)
    return result


def update_user(user_id, first_name, last_name, email, score, comment):
    """Updates a user's information in the database.

    :param user_id: The user's ID.
    :type user_id: int
    :param first_name: The user's first name.
    :type first_name: str
    :param last_name: The user's last name.
    :type last_name: str
    :param email: The user's email address (can be used as a user name).
    :type email: str
    :param score: The user's score from 0.0 to 100.0.
    :type score: float
    :param comment: Any additional comments.
    :type comment: str
    return: The number of rows affected. A value other than 1 indicates
        an error.
    :rtype: int
    """
    rows_affected = 0
    try:
        conn = connect()
        cursor = conn.cursor()
        sql = """ UPDATE User 
        SET FirstName = ?,
            LastName = ?,
            Email = ?,
            Score = ?,
            Comment = ?
        WHERE UserID = ?
        """
        data = (first_name, last_name, email, str(score), comment,
                str(user_id))
        cursor.execute(sql, data)
        conn.commit()
        rows_affected = cursor.rowcount
        cursor.close()
        conn.close()
    except Exception:
        ex = co.log_error(sys.exc_info())
        if co.DISPLAY_ERRORS:
            print(ex)
    return rows_affected


def delete_user(user_id):
    """Deletes a user from the database.

    :param user_id: The user's ID.
    :type user_id: int
    :returns: The number of rows affected. A value other than 1
        indicates an error.
    :rtype: int
    """
    rows_affected = 0
    try:
        conn = connect()
        cursor = conn.cursor()
        sql = "DELETE FROM User WHERE UserID = ?;"
        cursor.execute(sql, str(user_id))
        conn.commit()
        rows_affected = cursor.rowcount
        cursor.close()
        conn.close()
    except Exception:
        ex = co.log_error(sys.exc_info())
        if co.DISPLAY_ERRORS:
            print(ex)
    return rows_affected


def get_next_user_id():
    """Gets the anticipated value of the next UserID (usually the last row
        inserted) from the User table.

    :returns: The value of the next UserID or 0 if there is no data.
    :rtype: int
    """
    next_user_id = 0
    try:
        conn = connect()
        cursor = conn.cursor()
        sql = "SELECT MAX(UserID) as maxUserID FROM User;"
        cursor.execute(sql)
        row = cursor.fetchone()
        cursor.close()
        conn.close()
        next_user_id = (0 if row["maxUserID"] is
                        None else row["maxUserID"]) + 1
    except Exception:
        ex = co.log_error(sys.exc_info())
        if co.DISPLAY_ERRORS:
            print(ex)
    # Add 1 to the last user ID
    return next_user_id


def user_exists(email):
    """Checks if the given users exists in the database.
        Julen Pardo came up with this.
        Thought about changing the method to retrieve the UserID instead,
        but Email is supposed to be unique.
        If the count != 1, that means there are no users or more than one,
        which means something is wrong. This is a better method.

    :param email: The email to check.
    :type email: str
    :returns: True if the users exists, False if not.
    :rtype: bool
    """
    exists = False
    try:
        conn = connect()
        cursor = conn.cursor()
        sql = "SELECT COUNT(*) AS Count FROM User WHERE Email = ?;"
        cursor.execute(sql, email)
        result = cursor.fetchone()
        exists = (result["Count"] == 1)
        cursor.close()
        conn.close()
    except Exception:
        ex = co.log_error(sys.exc_info())
        if co.DISPLAY_ERRORS:
            print(ex)
    return exists


def connect():
    """Connects to the database.

    :returns: The connection to the SQLite database
    :rtype: object
    """
    conn = None
    try:
        conn = sqlite3.connect(PATH_TO_SQLITE_DB)
        # Use this to get objects by their column names
        # instead of their column indexes
        conn.row_factory = sqlite3.Row
    except Exception:
        ex = co.log_error(sys.exc_info())
        if co.DISPLAY_ERRORS:
            print(ex)
    return conn


def database_exists():
    """Create and populate the database if it does not exists.

    :return: True if the database exists or was created, false if not
    :rtype: bool
    """
    exists = True
    try:
        # Removed Pathlib (redundant) and needed to set PWD to correct
        # directory using os
        if not os.path.isdir(co.ROOT_DIR + os.path.sep + "data"):
            os.makedirs(co.ROOT_DIR + os.path.sep + "data", exist_ok=True)
        if not os.path.exists(PATH_TO_SQLITE_DB):
            if create_user_table() != -1:
                exists = False
            if create_user(
                "Rob", "Garcia", "rgarcia@rgprogramming.com",
                    80.0, "Administrator.") == 0:
                exists = False
            if create_user(
                "George", "Washington", "gwashington@rgprogramming.com",
                    90.0, "Old user.") == 0:
                exists = False
            if create_user(
                "Baby", "Yoda", "byoda@rgprogramming.com",
                    100.0, "New user.") == 0:
                exists = False
    # Use module_ex to prevent overwriting function internal exceptions (ex)
    except Exception:
        module_ex = co.log_error(sys.exc_info())
        if co.DISPLAY_ERRORS:
            print(module_ex)
    return exists

File: CSharp/Hello/Models/DatabaseFunctions.cs
/*
 * Handles all calls to the User database.
 * Do not instantiate; the methods in CommonFunctions should be accessed in a static way.
 *
 * .NET Core version used: 3.1.0
 * C# version used: 8.0
 * SQLite version used: 3.30.1
 *
 * Styling guide: .NET Core Engineering guidelines
 *     (https://github.com/dotnet/aspnetcore/wiki/Engineering-guidelines#coding-guidelines) and
 *     C# Programming Guide
 *     (https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/inside-a-program/coding-conventions)
 *
 * @category  CodersCompanion
 * @package   CSharp
 * @author    Rob Garcia <rgarcia@rgprogramming.com>
 * @license   https://opensource.org/licenses/MIT The MIT License
 * @link      https://github.com/garciart/CodersCompanion
 * @copyright 2019-2020 Rob Garcia
 */
using System.Data.SQLite;

namespace Hello.Models
{
    using System;
    using System.Data;
    using System.IO;

    /// <summary>
    /// Database Class. Handles all calls to the User database.
    /// Class methods:
    /// int CreateUserTable()
    /// long CreateUser(string firstName, string lastName, string email, float score, string comment)
    /// DataTable GetAllUsers()
    /// DataTable GetUserByUserID(long userID)
    /// DataTable GetUserByEmail(string email)
    /// int UpdateUser(long userID, string firstName, string lastName, string email, float score, string comment)
    /// int DeleteUser(long userID)
    /// long GetNextUserID()
    /// bool UserExists(string email)
    /// bool DatabaseExists()
    /// DatabaseFunctions()
    /// </summary>
    public class DatabaseFunctions
    {
        // Use readonly instead of const for PathToSQLiteDB, since it must be generated dynamically in C#
        private static readonly string PathToSQLiteDB = string.Format("{0}{1}Data{1}Users.db", CommonFunctions.RootDir, Path.DirectorySeparatorChar);

        /// <summary>
        /// Creates the User table if it does not exist in the database.
        /// </summary>
        /// <returns>The number of rows affected. While DROP will return -1, the last call, CREATE, should return 0.</returns>
        public static int CreateUserTable()
        {
            int rowsAffected = -1;
            try
            {
                using (SQLiteConnection conn = new SQLiteConnection(string.Format("URI=file:{0}", PathToSQLiteDB)))
                {
                    conn.Open();
                    string sql = @"DROP TABLE IF EXISTS User;";
                    using (SQLiteCommand cmd = new SQLiteCommand(sql, conn))
                    {
                        rowsAffected = cmd.ExecuteNonQuery();
                    }
                    sql = @"CREATE TABLE IF NOT EXISTS User (
                        UserID integer PRIMARY KEY,
                        FirstName text NOT NULL,
                        LastName text NOT NULL,
                        Email text UNIQUE NOT NULL,
                        Score real NOT NULL DEFAULT '100.0',
                        CreationDate text NOT NULL,
                        Comment text
                    );";
                    using (SQLiteCommand cmd = new SQLiteCommand(sql, conn))
                    {
                        rowsAffected = cmd.ExecuteNonQuery();
                    }
                }
            }
            catch (Exception ex)
            {
                string exception = CommonFunctions.LogError(ex);
                if (CommonFunctions.DisplayErrors) Console.WriteLine(exception);
            }
            return rowsAffected;
        }

        /// <summary>
        /// Inserts a new user into the database.
        /// </summary>
        /// <param name="firstName">The user's first name.</param>
        /// <param name="lastName">The user's last name.</param>
        /// <param name="email">The user's email address (can be used as a user name).</param>
        /// <param name="score">The user's score from 0.0 to 100.0.</param>
        /// <param name="comment">Any additional comments.</param>
        /// <returns>The rowid of the new user. A value of 0 indicates an error.</returns>
        public static long CreateUser(string firstName, string lastName, string email, float score, string comment)
        {
            long lastRowID = 0;
            try
            {
                using (SQLiteConnection conn = new SQLiteConnection(string.Format("URI=file:{0}", PathToSQLiteDB)))
                {
                    conn.Open();
                    string sql = @"INSERT INTO User VALUES (
                        @UserID,
                        @FirstName,
                        @LastName,
                        @Email,
                        @Score,
                        @CreationDate,
                        @Comment
                    );";
                    using (SQLiteCommand cmd = new SQLiteCommand(sql, conn))
                    {
                        string creationDate = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss");
                        cmd.Parameters.AddWithValue("@UserID", GetNextUserID());
                        cmd.Parameters.AddWithValue("@FirstName", firstName);
                        cmd.Parameters.AddWithValue("@LastName", lastName);
                        cmd.Parameters.AddWithValue("@Email", email);
                        cmd.Parameters.AddWithValue("@Score", score);
                        cmd.Parameters.AddWithValue("@CreationDate", creationDate);
                        cmd.Parameters.AddWithValue("@Comment", comment);
                        cmd.Prepare();
                        cmd.ExecuteNonQuery();
                        lastRowID = conn.LastInsertRowId;
                    }
                }
            }
            catch (Exception ex)
            {
                string exception = CommonFunctions.LogError(ex);
                if (CommonFunctions.DisplayErrors) Console.WriteLine(exception);
            }
            return lastRowID;
        }

        /// <summary>
        /// Gets all the users in the database and their information.
        /// </summary>
        /// <returns>
        /// An reader object of all the users in the database and their information. An empty object indicates an error.
        /// </returns>
        public static DataTable GetAllUsers()
        {
            // Must use DataTable to save the ResultSet after the connection closes
            DataTable result = new DataTable();
            try
            {
                using (SQLiteConnection conn = new SQLiteConnection(string.Format("URI=file:{0}", PathToSQLiteDB)))
                {
                    conn.Open();
                    string sql = @"SELECT *
                        FROM User
                        ORDER BY UserID ASC;";
                    using (SQLiteCommand cmd = new SQLiteCommand(sql, conn))
                    {
                        using (SQLiteDataReader reader = cmd.ExecuteReader(CommandBehavior.CloseConnection))
                        {
                            result.Load(reader);
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                string exception = CommonFunctions.LogError(ex);
                if (CommonFunctions.DisplayErrors) Console.WriteLine(exception);
            }
            return result;
        }

        /// <summary>
        /// Returns a single user and his or her information.
        /// </summary>
        /// <param name="userID">The user's ID.</param>
        /// <returns>
        /// The user's information indexed by column name or empty if the user's ID is not found.
        /// </returns>
        public static DataTable GetUserByUserID(long userID)
        {
            // Must use DataTable to save the ResultSet after the connection closes
            DataTable result = new DataTable();
            try
            {
                using (SQLiteConnection conn = new SQLiteConnection(string.Format("URI=file:{0}", PathToSQLiteDB)))
                {
                    conn.Open();
                    string sql = @"SELECT *
                        FROM User
                        WHERE UserID = @UserID;";
                    using (SQLiteCommand cmd = new SQLiteCommand(sql, conn))
                    {
                        cmd.Parameters.AddWithValue("@UserID", userID);
                        cmd.Prepare();
                        using (SQLiteDataReader reader = cmd.ExecuteReader(CommandBehavior.CloseConnection))
                        {
                            result.Load(reader);
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                string exception = CommonFunctions.LogError(ex);
                if (CommonFunctions.DisplayErrors) Console.WriteLine(exception);
            }
            return result;
        }

        /// <summary>
        /// Returns a single user and his or her information.
        /// </summary>
        /// <param name="email">The user's email.</param>
        /// <returns>
        /// The user's information indexed by column name or empty if the user's email is not found.
        /// </returns>
        public static DataTable GetUserByEmail(string email)
        {
            // Must use DataTable to save the ResultSet after the connection closes
            DataTable result = new DataTable();
            try
            {
                using (SQLiteConnection conn = new SQLiteConnection(string.Format("URI=file:{0}", PathToSQLiteDB)))
                {
                    conn.Open();
                    string sql = @"SELECT *
                    FROM User
                    WHERE Email = @Email;";
                    using (SQLiteCommand cmd = new SQLiteCommand(sql, conn))
                    {
                        cmd.Parameters.AddWithValue("@Email", email);
                        cmd.Prepare();
                        using (SQLiteDataReader reader = cmd.ExecuteReader(CommandBehavior.CloseConnection))
                        {
                            result.Load(reader);
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                string exception = CommonFunctions.LogError(ex);
                if (CommonFunctions.DisplayErrors) Console.WriteLine(exception);
            }
            return result;
        }

        /// <summary>
        /// Updates a user's information in the database.
        /// </summary>
        /// <param name="userID">The user's ID.</param>
        /// <param name="firstName">The user's first name.</param>
        /// <param name="lastName">The user's last name.</param>
        /// <param name="email">The user's email address (can be used as a user name).</param>
        /// <param name="score">The user's score from 0.0 to 100.0.</param>
        /// <param name="comment">Any additional comments.</param>
        /// <returns>
        /// The number of rows affected. A value other than 1 indicates an error.
        /// </returns>
        public static int UpdateUser(long userID, string firstName, string lastName, string email, float score, string comment)
        {
            int rowsAffected = 0;
            try
            {
                using (SQLiteConnection conn = new SQLiteConnection(string.Format("URI=file:{0}", PathToSQLiteDB)))
                {
                    conn.Open();
                    string sql = @"UPDATE User
                    SET FirstName = @FirstName,
                    LastName = @LastName,
                    Email = @Email,
                    Score = @Score,
                    Comment = @Comment
                    WHERE UserID = @UserID;";
                    using (SQLiteCommand cmd = new SQLiteCommand(sql, conn))
                    {
                        cmd.Parameters.AddWithValue("@FirstName", firstName);
                        cmd.Parameters.AddWithValue("@LastName", lastName);
                        cmd.Parameters.AddWithValue("@Email", email);
                        cmd.Parameters.AddWithValue("@Score", score);
                        cmd.Parameters.AddWithValue("@Comment", comment);
                        cmd.Parameters.AddWithValue("@UserID", userID);
                        cmd.Prepare();
                        rowsAffected = cmd.ExecuteNonQuery();
                    }
                }
            }
            catch (Exception ex)
            {
                string exception = CommonFunctions.LogError(ex);
                if (CommonFunctions.DisplayErrors) Console.WriteLine(exception);
            }
            return rowsAffected;
        }

        /// <summary>
        /// Deletes a user from the database.
        /// </summary>
        /// <param name="userID">The user's ID.</param>
        /// <returns>
        /// The number of rows affected. A value other than 1 indicates an error.
        /// </returns>
        public static int DeleteUser(long userID)
        {
            int rowsAffected = 0;
            try
            {
                using (SQLiteConnection conn = new SQLiteConnection(string.Format("URI=file:{0}", PathToSQLiteDB)))
                {
                    conn.Open();
                    string sql = @"DELETE FROM User WHERE UserID = @UserID;";
                    using (SQLiteCommand cmd = new SQLiteCommand(sql, conn))
                    {
                        cmd.Parameters.AddWithValue("@UserID", userID);
                        cmd.Prepare();
                        rowsAffected = cmd.ExecuteNonQuery();
                    }
                }
            }
            catch (Exception ex)
            {
                string exception = CommonFunctions.LogError(ex);
                if (CommonFunctions.DisplayErrors) Console.WriteLine(exception);
            }
            return rowsAffected;
        }

        /// <summary>
        /// Gets the anticipated value of the next UserID (usually the last row
        /// inserted) from the User table.
        /// </summary>
        /// <returns>The value of the next UserID or 0 if there is no data.</returns>
        public static long GetNextUserID()
        {
            long nextUserID = 0;
            try
            {
                using (SQLiteConnection conn = new SQLiteConnection(string.Format("URI=file:{0}", PathToSQLiteDB)))
                {
                    conn.Open();
                    string sql = @"SELECT MAX(UserID) as maxUserID FROM User;";
                    using (SQLiteCommand cmd = new SQLiteCommand(sql, conn))
                    {
                        object o = cmd.ExecuteScalar();
                        nextUserID = (o is System.DBNull ? 0 : Convert.ToInt64(o)) + 1;
                    }
                }
            }
            catch (Exception ex)
            {
                string exception = CommonFunctions.LogError(ex);
                if (CommonFunctions.DisplayErrors) Console.WriteLine(exception);
            }
            return nextUserID;
        }

        /// <summary>
        /// Checks if the given users exists in the database.
        /// Julen Pardo came up with this.
        /// Thought about changing the method to retrieve the UserID instead,
        /// but Email is supposed to be unique.
        /// If the count != 1, that means there are no users or more than one,
        /// which means something is wrong. This is a better method.
        /// </summary>
        /// <param name="email">The email to check.</param>
        /// <returns>True if the users exists, false if not.</returns>
        public static bool UserExists(string email)
        {
            bool exists = false;
            try
            {
                using (SQLiteConnection conn = new SQLiteConnection(string.Format("URI=file:{0}", PathToSQLiteDB)))
                {
                    conn.Open();
                    string sql = @"SELECT COUNT(*) AS Count
                        FROM User
                        WHERE Email = @Email;";
                    using (SQLiteCommand cmd = new SQLiteCommand(sql, conn))
                    {
                        cmd.Parameters.AddWithValue("@Email", email);
                        cmd.Prepare();
                        long count = (long)cmd.ExecuteScalar();
                        exists = count == 1 ? true : false;
                    }
                }
            }
            catch (Exception ex)
            {
                string exception = CommonFunctions.LogError(ex);
                if (CommonFunctions.DisplayErrors) Console.WriteLine(exception);
            }
            return exists;
        }

        /// <summary>
        /// Creates and populates the database if it does not exist.
        /// </summary>
        /// <returns>True if the database exists or was create, false if not.</returns>
        public static bool DatabaseExists()
        {
            bool exists = true;
            try
            {
                Directory.SetCurrentDirectory(CommonFunctions.ModelDir);
                string dbFolder = string.Format("{0}{1}Data", CommonFunctions.RootDir, Path.DirectorySeparatorChar);
                string dbFile = PathToSQLiteDB;
                if (!Directory.Exists(dbFolder) || !File.Exists(dbFile))
                {
                    // Creates the Data directory if it does not exist
                    Directory.CreateDirectory(dbFolder);
                    // Creates the database file if it does not exist
                    if (CreateUserTable() != 0) exists = false;
                    // Set initial values
                    if (CreateUser("Rob", "Garcia", "rgarcia@rgprogramming.com", 80.0f, "Administrator.") == 0) exists = false;
                    if (CreateUser("Thomas", "Jefferson", "tjefferson@rgprogramming.com", 90.0f, "Old user.") == 0) exists = false;
                    if (CreateUser("Baby", "Yoda", "byoda@rgprogramming.com", 100.0f, "New user.") == 0) exists = false;
                }
            }
            catch (Exception ex)
            {
                string exception = CommonFunctions.LogError(ex);
                if (CommonFunctions.DisplayErrors) Console.WriteLine(exception);
            }
            return exists;
        }

        /// <summary>
        /// Constructor.
        /// </summary>
        public DatabaseFunctions()
        {
            Console.WriteLine("Do not instantiate. The methods in DatabaseFunctions should be accessed in a static way.");
            System.Environment.Exit(-1);
        }
    }
}